src/protocols/secure_channel/CASEServer.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <protocols/secure_channel/CASEServer.h>
+
+#include <core/CHIPError.h>
+#include <support/CodeUtils.h>
+#include <support/SafeInt.h>
+#include <transport/SecureSessionMgr.h>
+
+using namespace ::chip::Inet;
+using namespace ::chip::Transport;
+
+namespace chip {
+
+CHIP_ERROR CASEServer::ListenForSessionEstablishment(Messaging::ExchangeManager * exchangeManager, TransportMgrBase * transportMgr,
+                                                     SecureSessionMgr * sessionMgr, Transport::AdminPairingTable * admins)
+{
+    VerifyOrReturnError(transportMgr != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(sessionMgr != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(admins != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mSessionMgr      = sessionMgr;
+    mAdmins          = admins;
+    mExchangeManager = exchangeManager;
+
+    ReturnErrorOnFailure(mPairingSession.MessageDispatch().Init(transportMgr));
+
+    ReturnErrorOnFailure(
+        mExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::CASE_SigmaR1, this));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
+{
+    ReturnErrorCodeIf(ec == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Lookup the admin that corresponds to the CASE session setup request.
+    // Each admin provisions their own credentials on the device. So it's essential to
+    // use the correct operational certificates for CASE session setup.
+    mAdminId = ec->GetSecureSession().GetAdminId();
```

Comment:
Indexing and lookup up by Admin ID breaks scoping.  Per the spec, the admin must be located based upon the root PK of the root CA for the requesting agent.

Reply:
That makes sense. We can do that as part of multi admin feature.
Would you prefer if I updated the above comment as part of this PR?

Reply:
A TODO seems appropriate.  It doesn't have to happen now.  I'm just pointing out that this code does not appear to be compliant with the `4.369 Validate Sigma1` algorithm.

Reply:
Updated the TODO here: https://github.com/project-chip/connectedhomeip/pull/6810/files#diff-c9e9c579ef5bbdbb97683c09136f253a842efdbb8d8febaf3e92ca12b47a0dddR56